File: 3/day3.go
package day3

import (
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type Location struct {
	line   int
	start  int
	stop   int
	number int
}

func newLocation(lineno int, start_idx int, end_idx int, number int) Location {
	loc := Location{}

	loc.line = lineno
	loc.start = start_idx
	loc.stop = end_idx
	loc.number = number

	return loc
}

func getRune(line int, char int, data []string) rune {
	if line < 0 || line >= len(data) {
		return '.'
	}

	row := data[line]

	if char < 0 || char >= len(row) {
		return '.'
	}

	return rune(row[char])
}

func (self Location) pointsToCheck(data string) []rune {
	rows := strings.Split(data, "\n")

	runes := make([]rune, 0, 10)
	for c := self.start - 1; c <= self.stop+1; c++ {

		for l := self.line - 1; l <= self.line+1; l++ {
			// line before, current line, next line
			rune := getRune(l, c, rows)
			runes = append(runes, rune)
		}

	}
	return runes
}

func is_digit(char rune) bool {
	return char >= '0' && char <= '9'
}

func to_int(char string) int {
	value, err := strconv.Atoi(char)
	check(err)
	return value
}

func idenitfy_numbers(input string) []Location {
	rows := strings.Split(input, "\n")

	// 1. identify numbers with indexes
	// -> 467 at 0:0-3, 114 at ...
	// 2. identify characters around numbers
	// 467: ...*. etc

	// number: line, start, stop

	numbers := make([]Location, 0, 10)

	for lineno, row := range rows {
		open_number := false
		current_number := 0
		start_idx := -1
		end_idx := -1

		for charno, char := range row {
			end_idx = charno
			if is_digit(char) {
				current_number = current_number*10 + to_int(string(char))

				if !open_number {
					open_number = true
					start_idx = charno
				}

			} else if open_number {
				// numbers[current_number] = newLocation(lineno, start_idx, end_idx, current_number)
				numbers = append(numbers, newLocation(lineno, start_idx, end_idx-1, current_number))

				current_number = 0
				open_number = false
				start_idx = -1
			}
		}
		// number touching right edge:
		if open_number {
			numbers = append(numbers, newLocation(lineno, start_idx, end_idx, current_number))
		}
	}

	return numbers
}

// func getSymbols() []rune {
// 	runes := make([]rune, 5, 5)

// 	runes[0] = '*'
// 	runes[1] = '#'
// 	runes[3] = '+'
// 	runes[4] = '$'

// 	return runes
// }

func charsToStrings(data []rune) []string {
	strings := make([]string, len(data))

	for idx, char := range data {
		strings[idx] = string(char)
	}

	return strings
}

func checkLocation(loc Location, rows string) int {
	chars := loc.pointsToCheck(rows)

	for _, char := range chars {

		if !is_digit(char) && char != '.' {
			return loc.number
		}
	}

	return 0
}

func run(filename string) int {
	_dat, err := os.ReadFile(filename)
	check(err)

	dat := string(_dat)

	numbers := idenitfy_numbers(dat)

	result := 0
	for _, location := range numbers {
		result += checkLocation(location, dat)
	}

	return result
}

File: 3/day3.rs
use std::fs;


#[derive(Debug)]
struct Position {
    line: i32,
    start: i32,
    end: i32,
    num: u32,
}

fn len<T>(something: &Vec<T>) -> i32 {
    return something.len() as i32;
}

impl Position {
    // Define the get_score function for the Position struct
    fn is_valid(&self, rows: Vec<&str>) -> bool {
        for _y in 0..=2 {
            let y = self.line + (_y - 1);

            if y < 0 || y >= len(&rows) {
                continue;
            }

            let row = rows[y as usize].chars().collect();

            for x in self.start - 1..=self.end + 1 {
                if x < 0 || x >= len(&row) {
                    continue;
                }

                let value = row[x as usize];

                if is_digit(value) || value == '.' {
                    continue;
                }

                // else: it's a symbol so this Position is valid!
                return true;
            }
        }


        return false;
    }

    fn get_score(&self, rows: Vec<&str>) -> u32 {
        if self.is_valid(rows) {
            return self.num;
        }

        return 0; // because * 1
    }
}

fn is_digit(char: char) -> bool {
    return char >= '0' && char <= '9';
}

fn find_numbers(line: &str, lineno: u32, collect: &mut Vec<Position>) {
    let mut current_number = 0;

    let mut start_idx = -1;
    let mut end_idx = -1;

    for (idx, char) in line.chars().enumerate() {
        end_idx = idx as i32;
        if is_digit(char) {
            if current_number == 0 {
                start_idx = idx as i32;
            }

            current_number = current_number * 10 + char.to_digit(10).unwrap()
        } else {
            if current_number != 0 {
                collect.push(Position { line: lineno as i32, start: start_idx, end: end_idx - 1, num: current_number });
            }

            current_number = 0;
        }
    }

    if current_number != 0 {
        collect.push(Position { line: lineno as i32, start: start_idx, end: end_idx, num: current_number });
    }
}

fn run(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let rows: Vec<&str> = contents.split("\n").collect();

    let mut numbers: Vec<Position> = vec![];

    for (idx, line) in rows.iter().enumerate() {
        find_numbers(line, idx as u32, &mut numbers)
    }

    let mut result = 0;

    for pos in &numbers {
        let _result = pos.get_score(rows.clone());
        result += _result;
    }

    return result;
}

fn process_symbols(rows: Vec<&str>, numbers: Vec<Position>) -> u32 {
    // find all symbols
    // find all overlapping numbers
    // calculate
    // ...
    // profit!
    let mut result = 0;

    for (_lineno, row) in rows.iter().enumerate() {
        for (_charno, char) in row.chars().enumerate() {
            if is_digit(char) || char == '.' {
                // not a symbol
                continue;
            }
            let mut nums_found: Vec<&Position> = vec![];


            let lineno = _lineno as i32;
            let charno = _charno as i32;

            for position in &numbers {
                let matches_line = (position.line - 1..=position.line + 1).contains(&lineno);
                let matches_col = (position.start - 1..=position.start + 1).contains(&charno) || (position.end - 1..=position.end + 1).contains(&charno);


                if (matches_line && matches_col) {
                    nums_found.push(position);
                }
            }

            if nums_found.len() == 2 {
                // right amount of numbers!
                result += nums_found[0].num * nums_found[1].num;
            }
        }
    }

    return result;
}

fn run2(filename: &str) -> u32 {
    let contents = fs::read_to_string(filename)
        .expect("Should have been able to read the file");

    let rows: Vec<&str> = contents.split("\n").collect();

    let mut numbers: Vec<Position> = vec![];

    for (idx, line) in rows.iter().enumerate() {
        find_numbers(line, idx as u32, &mut numbers)
    }

    return process_symbols(rows, numbers);
}


fn main() {
    // let result = run("dev.txt");
    // println!("result: {r}", r = result);
    // assert!(result == 4361);
    //
    // let pt1_result = run("test.txt");
    // println!("part 1: {}", pt1_result);
    // assert!(pt1_result == 546563);
    //
    // println!("All checks passed for pt1!")

    let result = run2("dev.txt");
    println!("result: {}", result);
    assert!(result == 467835);

    println!("All checks passed for pt2!");

    let pt2_result = run2("test.txt");
    println!("part 2: {}", pt2_result);
}
